src/ast/passes/semantic_analyser.cpp

```diff
@@ -2082,6 +2123,35 @@ void SemanticAnalyser::visit(For &f)
   if (has_error())
     return;
 
+  // Collect a list of unique variables which are referenced in the loop's body
+  // and declared before the loop. These will be passed into the loop callback
+  // function as the context parameter.
+  CollectNodes<Variable> vars_referenced;
+  std::unordered_set<std::string> var_set;
+  for (auto *stmt : *f.stmts) {
```

Comment:
Nit: looks like you could combine these two loops to get referenced and new vars.

Reply:
I'd like to delete the "new_vars" loop in hopefully not too long (when #3017 is done), so was keeping them separate to make that easier

Reply:
Could always add a comment todo with a reference to the issue but that's minor. I won't block.

Reply:
The comment on the next for-loop does reference the issue